File: flightservice/src/main/java/com/sebastiangds/services/EnvHelper.java
package com.sebastiangds.services;

import com.sebastiangds.flightservice.backendconnector.EndpointFactory;
import logging.SILevel;
import logging.Sender;

import java.util.HashMap;
import java.util.Map;

public class EnvHelper {
    private final String ENV_VAL_KEY = "ENV_TYPE";

    private EnvType envType = EnvType.DEV;
    private final Map<Service, Map<EnvType, ServiceInfo>> serviceMap = new HashMap<>();

    public EnvHelper() {
        populateMap();
    }

    private void setEnvType() {
        Map<String, String> sysEnv = System.getenv();
        if (sysEnv.containsKey(this.ENV_VAL_KEY)) {
            String envVar = sysEnv.get(this.ENV_VAL_KEY);
            try {
                this.envType = EnvType.valueOf(envVar);
            } catch (IllegalArgumentException e) {
                String warningMsg =
                        "Environment " + envVar + " provided by "
                                + this.ENV_VAL_KEY + " not recognised. " +
                                "Continue in " + envType.toString();
                System.out.println(this.ENV_VAL_KEY + " not recognised. " + warningMsg);
                /*Sender sender = new Sender(this.serviceMap.get(Service.RABBIT).get(this.envType).getHost());
                sender.makeLog(
                        EnvHelper.class.getName(),
                        SILevel.WARNING,
                        this.ENV_VAL_KEY + " not recognised",
                        warningMsg);*/
            }
        } else {
            /*Sender sender = new Sender(this.serviceMap.get(Service.RABBIT).get(this.envType).getHost());
            sender.makeLog(
                    EnvHelper.class.getName(),
                    SILevel.WARNING,
                    this.ENV_VAL_KEY + " not set",
                    "Defaulting to " + envType.toString());*/
            System.out.println(this.ENV_VAL_KEY + " not set. " +
                    "Defaulting to " + envType.toString());
        }
    }

    private void populateMap () {
        Map<EnvType, ServiceInfo> feedbackMap = new HashMap<>();
        feedbackMap.put(EnvType.DEV, new ServiceInfo("localhost", 3333));
        feedbackMap.put(EnvType.COMPOSE, new ServiceInfo("feedback-service", 3333));
        this.serviceMap.put(Service.FEEDBACK, feedbackMap);

        Map<EnvType, ServiceInfo> foodMap = new HashMap<>();
        foodMap.put(EnvType.DEV, new ServiceInfo("localhost", 5009));
        foodMap.put(EnvType.COMPOSE, new ServiceInfo("food-service", 5009));
        this.serviceMap.put(Service.FOOD, foodMap);

        Map<EnvType, ServiceInfo> flightMap = new HashMap<>();
        flightMap.put(EnvType.DEV, new ServiceInfo("localhost", 6565));
        flightMap.put(EnvType.COMPOSE, new ServiceInfo("flight-service", 6565));
        this.serviceMap.put(Service.FLIGHT, flightMap);

        Map<EnvType, ServiceInfo> mongoMap = new HashMap<>();
        mongoMap.put(EnvType.DEV, new ServiceInfo("localhost", 27017));
        mongoMap.put(EnvType.COMPOSE, new ServiceInfo("mongodb", 27017));
        this.serviceMap.put(Service.MONGO, mongoMap);

        Map<EnvType, ServiceInfo> rabbitMap = new HashMap<>();
        rabbitMap.put(EnvType.DEV, new ServiceInfo("localhost", 5672, "guest", "guest"));
        rabbitMap.put(EnvType.COMPOSE, new ServiceInfo("rabbitmq", 5672, "guest", "guest"));
        this.serviceMap.put(Service.RABBIT, rabbitMap);
    }

    public ServiceInfo getService(Service serviceName) {
        setEnvType();
        /*Sender sender = new Sender(this.serviceMap.get(Service.RABBIT).get(this.envType).getHost());*/
        ServiceInfo info = this.serviceMap.get(serviceName).get(this.envType);
        /*sender.makeLog(EnvHelper.class.getName(), SILevel.INFO,
                "Retrieved service info for " + serviceName.toString(),
                "Info was host <" + info.getHost() + "> port <" + info.getPort() + ">");*/
        return info;
    }

    public EnvType getEnvType() {
        return this.envType;
    }
}

File: feedback/env_helper.js
const process = require('process');

const EnvType = Object.freeze({
  "DEV":"DEV",
  "COMPOSE":"COMPOSE",
  "TEST":"TEST",
  "PROD":"PROD",
});

const envValKey = "ENV_TYPE";
let envType = EnvType.DEV;
const serviceMap = {};

const service = Object.freeze({
  "FOOD":"FOOD",
  "FEEDBACK":"FEEDBACK",
  "RABBIT":"RABBIT",
  "FLIGHT":"FLIGHT",
  "MONGO":"MONGO",
});

class ServiceInfo {
  constructor(host, port, userName, password) {
    this.host = host;
    this.port = port;
    this.userName = userName;
    this.password = password;
  }
}

function getService(serviceName) {
  const info = serviceMap[serviceName][envType];
  console.log(
    "Retrieved service info for " + serviceName.toString(),
    "Info was host " + info.host + ":" + info.port
  );
  return info;
}


function setUp() {
  let envStr = process.env[envValKey];
  if (envStr) {

    envStr = envStr.toString().toUpperCase();

    if (EnvType[envStr]) {
      envType = EnvType[envStr];
    } else {

      const warningMsg = 
        "Environment " + envStr + " provided by "
        + envValKey + " not recognised. " +
        "Continue in " + envType;
      console.log(envValKey + " not recognised. " + warningMsg);
    }
  } else {
    console.log(
      envValKey + " not set. ",
      "Defaulting to " + envType
    );
  }
  populateMap();
}

function populateMap() {
  const feedbackMap = {};
  feedbackMap[EnvType.DEV] = new ServiceInfo("localhost", 3333);
  feedbackMap[EnvType.COMPOSE] = new ServiceInfo("feedback-service", 3333);
  serviceMap[service.FEEDBACK] = feedbackMap;

  const foodMap = {};
  foodMap[EnvType.DEV] = new ServiceInfo("localhost", 5009);
  foodMap[EnvType.COMPOSE] = new ServiceInfo("food-service", 5009);
  serviceMap[service.FOOD] = foodMap;

  const flightMap = {};
  flightMap[EnvType.DEV] = new ServiceInfo("localhost", 6565);
  flightMap[EnvType.COMPOSE] = new ServiceInfo("flight-service", 6565);
  serviceMap[service.FLIGHT] = flightMap;

  const mongoMap = {};
  mongoMap[EnvType.DEV] = new ServiceInfo("localhost", 27017);
  mongoMap[EnvType.COMPOSE] = new ServiceInfo("mongodb", 27017);
  serviceMap[service.MONGO] = mongoMap;

  const rabbitMap = {};
  rabbitMap[EnvType.DEV] = new ServiceInfo("localhost", 5672, "guest", "guest");
  rabbitMap[EnvType.COMPOSE] = new ServiceInfo("rabbitmq", 5672, "guest", "guest");
  serviceMap[service.RABBIT] = rabbitMap;
}

setUp();

exports.envType = envType;
exports.EnvType = EnvType;
exports.service = service;
exports.getService = getService;
exports.ServiceInfo = ServiceInfo;